src/rpc/blockchain.cpp

```diff
@@ -552,15 +570,23 @@ static RPCHelpMan getblockheader()
         throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Block not found");
     }
 
-    if (!fVerbose)
+    UniValue headers{UniValue::VARR};
+    do
     {
-        CDataStream ssBlock(SER_NETWORK, PROTOCOL_VERSION);
-        ssBlock << pblockindex->GetBlockHeader();
-        std::string strHex = HexStr(ssBlock);
-        return strHex;
+        if (!fVerbose)
+        {
+            CDataStream ssBlock(SER_NETWORK, PROTOCOL_VERSION);
+            ssBlock << pblockindex->GetBlockHeader();
+            headers.push_back(HexStr(ssBlock));
+        }
+        else
+            headers.push_back(blockheaderToJSON(tip, pblockindex));
+
+        pblockindex = chainman.ActiveChain().Next(pblockindex);
```

Comment:
I'm pretty sure calling `Next()` requires the thread to hold `cs_main`. But I don't think you want to grab and release the lock on each iteration of this loop; it's probably better to take the lock close to where you currently do, and hold it throughout this loop.

But it's not ideal to hold `cs_main` during the conversions to JSON or (for non-verbose) encodings to hex. So you may want to consider creating a vector of `CBlockindex` pointers, have each iteration of this loop `push_back()` to this vector,  drop `cs_main`, then run a  separate loop to format the results. (This is okay because once you have a `CBlockIndex` pointer, both the pointer and the object it points to remain valid without having to hold `cs_main`.)

Reply:
It appears to be what caused CI to fail with `-Wthread-safety-analysis` enabled.

Reply:
I have implemented @LarryRuane's suggestion to store the `CBlockindex` pointers and then iterate through them. However, the CI failure is still occurring. From what I recall, this wasn't happening before. Any insight into why this is now the case?